clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/ChainBuilder.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.client.internal.store;
+
+import org.ehcache.clustered.common.store.Chain;
+import org.ehcache.clustered.common.store.Element;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Builds {@link Chain}s
+ */
+public class ChainBuilder {
+
+  private List<Element> elements = new ArrayList<Element>();
+
+  public ChainBuilder() {
+
+  }
+
+  private ChainBuilder(List<Element> elements) {
+    this.elements = elements;
+  }
+
+  public ChainBuilder add(final ByteBuffer payload) {
+    List<Element> newList = new ArrayList<Element>();
+    newList.addAll(this.elements);
+    newList.add(new Element() {
```

Comment:
@rkavanap : this is our standard builder pattern: the builder itself is immutable and always returns a modified copy of itself. though I agree that since this builder isn't meant for end-user consumption, many array allocations and copies might be overkill so we may consider using a mutable builder in this case.